base/essentials.jl

```diff
@@ -312,10 +312,34 @@ struct Colon
 end
 const (:) = Colon()
 
-# For passing constants through type inference
-struct Val{T}
+"""
+    Val(c)
+
+Constructs a `Val{c}()`, which contains no run-time data. Types like this can be used to
+pass the information between functions through the value `c`, which must be an `isbits`
+value. The intent of this construct is to be able to dispatch on constants directly (at
+compile time) without having test the value of the constant at run time.
+
+### Example
+
+```jldoctest
+julia> f(::Val{true}) = "Good"
+f (generic function with 1 method)
+
+julia> f(::Val{false}) = "Bad"
+f (generic function with 2 methods)
+
+julia> f(Val(true))
+"Good"
+```
+"""
+struct Val{x}
 end
 
+Val(x) = (@_pure_meta; Val{x}())
+
+show(io::IO, ::Val{x}) where {x} = print(io, "Val($x)")
```

Comment:
The `RefValue` example exposes larger differences, e.g. a concrete type, and we allow `RefValue{Any}(0)` also. In the case of `Val` the relation is one-to-one; I wouldn't have done it otherwise.